ofrak_core/ofrak/gui/script_builder.py

```diff
@@ -0,0 +1,297 @@
+
+        if False:
+            import ofrak_binary_ninja
+            import ofrak_capstone
+            ofrak.discover(ofrak_capstone)
+            ofrak.discover(ofrak_binary_ninja)
+
+        if False:
+            import ofrak_ghidra
+            ofrak.discover(ofrak_ghidra)
+
+        ofrak = OFRAK()
+        ofrak.run(main)
+    """
+
+
+class ScriptBuilder:
+    """
+    Builds and maintains runnable OFRAK scripts as sequences of actions, with each script tied to
+    a session.
+    """
+
+    def __init__(self):
+        self.root_cache: Dict[bytes, Resource] = {}
+        self.script_sessions: Dict[bytes, ScriptSession] = {}
+        self.selectable_indexes: List[ResourceIndexedAttribute] = [
+            FilesystemEntry.Name,
+            Data.Offset,
+        ]
+
+    async def add_action(
+        self,
+        resource: Resource,
+        action: str,
+        action_type: ActionType,
+    ) -> None:
+        """
+        :param action:
+        :param action_type:
+        """
+        var_name = await self.add_variable(resource)
+        qualified_action = action.format(resource=var_name)
+        await self._add_action_to_session(resource, qualified_action, action_type)
+
+    async def add_variable(self, resource: Resource) -> bytes:
+        if await self._var_exists(resource):
+            return await self._get_variable_from_session(resource)
+
+        root_resource = await self._get_root_resource(resource)
+        if resource.get_id() == root_resource.get_id():
+            await self._add_variable_to_session(resource, "root_resource")
+            return "root_resource"
+
+        parent = await resource.get_parent()
+        if not await self._var_exists(parent):
+            await self.add_variable(parent)
+
+        selector = await self._get_selector(resource)
+        name = await self._generate_name(resource)
+        await self._add_action_to_session(
+            resource,
+            rf"""
+        {name} = {selector}""",
+            ActionType.UNDEF,
+        )
+        await self._add_variable_to_session(resource, name)
+        return name
+
+    async def delete_action(self, resource: Resource, action: str) -> None:
+        """
+        :param action:
+        """
+        root_resource = await self._get_root_resource(resource)
+        session = self._get_session(root_resource.get_id())
+        # TODO: do we really need to delete an action from the script?
+        for idx, script_action in enumerate(session.actions):
+            if script_action.action == action:
+                session.actions.pop(idx)
+
+    async def get_script(self, resource: Resource) -> str:
+        """
+        :return script:
+        """
+        root_resource = await self._get_root_resource(resource)
+        return self._get_script(root_resource.get_id())
+
+    async def get_all_of_type(self, resource: Resource, target_type: ActionType) -> str:
+        """
+        :param target_type:
+        :return script:
```

Comment:
In my opinion, these should either be filled in, or removed. Same goes for all instances of the parameters in other doc comments throughout the PR.